src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,531 @@
+                    console.log(
+                        `Computed reduced output: forkGenesisTimestamp=${reducedOutput.forkGenesisTimestamp}`
+                    );
+
+                    // Move to next fork
+                    const nextForkId =
+                        reducedOutput.latestBlock.stateSnapshotHash;
+                    currentForkId = nextForkId;
+                    isDisputed =
+                        await stateManager.stateChannelManagerContract.isForkDisputed(
+                            channelId,
+                            currentForkId
+                        );
+                } else {
+                    // No disputes provided, can't proceed
+                    return {
+                        isValid: false,
+                        error: "No disputes provided for disputed fork"
+                    };
+                }
+            }
+
+            // Get the fork genesis snapshot from the participant's payload
+            const participantProvidedSnapshot =
+                snapshotPayload.latestForkGenesisSnapshot;
+
+            // Verify that the computed fork ID matches the participant's claimed fork ID
+            if (currentForkId !== participantProvidedSnapshot.forkId) {
+                return {
+                    isValid: false,
+                    error: `Computed fork ID ${currentForkId} does not match participant provided fork ID ${participantProvidedSnapshot.forkId}`
+                };
+            }
+
+            console.log(`Verified fork ID matches: ${currentForkId}`);
+
+            // Verify state proof to prove the latest state from the proven fork
+            if (
+                snapshotPayload.stateProof &&
+                snapshotPayload.stateProof.milestones.length > 0
+            ) {
+                console.log(`Verifying state proof from proven fork`);
+
+                // Get the latest state proof from the proven fork to compare with participant's state proof
+                const latestBlockHeight =
+                    stateManager.storage.blocks.getNextBlockHeight(
+                        currentForkId
+                    ) - 1;
+                const computedStateProof =
+                    await stateManager.agreementManager.getStateProof(
+                        currentForkId,
+                        latestBlockHeight
+                    );
+
+                // Compare the state proof objects
+                if (!isEqual(computedStateProof, snapshotPayload.stateProof)) {
+                    return {
+                        isValid: false,
+                        error: "State proof does not match computed state proof"
+                    };
+                }
+
+                console.log(
+                    `Verified state proof matches computed state proof`
+                );
+
+                // Verify the encoded state by generating a state proof from it and comparing
+                if (snapshotPayload.encodedState) {
```

Comment:
don't understand the comment above this line of code - to verify the encoded state you just hash it and compare it to the latestVerifiedSnpashot from stateProof after stateProof is verified - if stateProof is empty it should be the genesisSnapshot and than the snapshot has a field `snapshotData.stateMachineStateHash` to which you compare the hash